### PR TITLE
[cryptotest] Add NIST AES-GCM testvector parser

### PR DESCRIPTION
**Pull request by sadmansakib-git**
This PR pulls in the NIST testvectors for AES-GCM, and includes the associated parser and JSON schema.

### CODE REVIEW

File: sw/host/cryptotest/testvectors/data/schemas/aes_gcm_schema.json [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        "maximum": 255
+      }
+    }
+  },
+  "type": "array",
+  "items": {
+    "type": "object",
+    "additionalProperties": false,
+    "properties": {
+      "test_case_id": {
```
Comment (RyanTorok) [inline sw/host/cryptotest/testvectors/data/schemas/aes_gcm_schema.json:21]: Nit: Add a `vendor` field here, to match the other cryptotest schemas.

File: sw/host/cryptotest/testvectors/parsers/cryptotest_util.py [hunk 2]
```diff
@@ -51,6 +51,75 @@ def parse_rsp(file_path: str) -> dict:
     return result
 
 
+def parse_gcm_rsp(file_path: str) -> dict:
```
[the remark this answers is not in the document]
Comment (sadmansakib-git) [reply sw/host/cryptotest/testvectors/parsers/cryptotest_util.py:54]: Let me explain what happens if we use the existing `parse_rsp()` handler with an example file:

1) Say, we are working with one of the testvectors from the zip file named `gcmEncryptExtIV128.rsp`. It contains a total of `7875` test vectors (we can extract the file from the zip and look for "Count" instances)

2) We can write a simple python script that will parse this `.rsp` file with `parse_rsp()` handler and dump the contents as a json file:

```python
import json

from cryptotest_util import parse_rsp


with open("demo.out.txt", "w") as fw:
    json.dump(parse_rsp("gcmEncryptExtIV128.rsp"), fw, indent=4)
```

3) Now we can verify whether we got all the contents from `.rsp` file or not.

If you perform the above steps one by one, you will observe that only `105` testvectors will be parsed and the rest will be gone. The reason behind loosing testvectors with `parse_rsp()` handler is the way it handles each testvectors. It creates a python dict with key equals to `curr_section_name`, but for gcm rsp files it is not unique. So everytime it overrides all the existing contents in the dict for duplicate keys.

Example:

A sample rsp file consisting of two test vectors:

```
[Keylen = 128]
[IVlen = 96]
[PTlen = 408]
[AADlen = 160]
[Taglen = 32]

Count = 0
Key = e64903a77d2c8f54e5741354895f9f25
IV = 75bfc0f3c6ac071af0434318
PT = 19561f57d57d9a961bbc6ac5634556d005fa6010d90bd218c6277537a43f8d3fa8f29a16e4cc495b49b8af195d917cb760c34f
AAD = 416b40f14bdb9f0acef996c963d23bcf10b72518
CT = 898abb3d7069c05919046fe48ca9a443a5d2bd2d28503fd0a2716b2ef5a175f74868f7917f5542144b6704df8a42cc11c965c3
Tag = 5c526f9d

[Keylen = 128]
[IVlen = 96]
[PTlen = 408]
[AADlen = 128]
[Taglen = 32]

Count = 0
Key = c692470a8a7d3fefea6901db64f2904e
IV = de3f853639e495eedb58e267
PT = 99132df39de42d3fab3f3291ed9a78b963b5efc70f5da13e05a8de9069763542ca66848e521fc8e27b0bff4c51a55ca5247740
AAD = 28423be900993a0b594e5ad71c6fff18
CT = 4e7eeb7853c2588d8a3af3283698c70ce2f360f3ad587145cf68ae8b3601d2ced5bf04c2807c7fd5222344883fd08e64b89d51
Tag = 668fb277
```

Output:

```
{
    "Taglen = 32": [
        {
            "Count": "0",
            "Key": "c692470a8a7d3fefea6901db64f2904e",
            "IV": "de3f853639e495eedb58e267",
            "PT": "99132df39de42d3fab3f3291ed9a78b963b5efc70f5da13e05a8de9069763542ca66848e521fc8e27b0bff4c51a55ca5247740",
            "AAD": "28423be900993a0b594e5ad71c6fff18",
            "CT": "4e7eeb7853c2588d8a3af3283698c70ce2f360f3ad587145cf68ae8b3601d2ced5bf04c2807c7fd5222344883fd08e64b89d51",
            "Tag": "668fb277"
        }
    ]
}
```

`parse_rsp()` handler will select `Taglen = 32` as key, resulting losing the first test vector.

I hope it makes sense now.

File: sw/host/cryptotest/testvectors/data/schemas/aes_gcm_schema.json (hunk 1 above)
Comment (sadmansakib-git) [reply sw/host/cryptotest/testvectors/data/schemas/aes_gcm_schema.json:25]: Done

File: sw/host/cryptotest/testvectors/parsers/cryptotest_util.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RyanTorok) [reply sw/host/cryptotest/testvectors/parsers/cryptotest_util.py:54]: After some offline conversation with @milesdai , I'm going to push a PR to migrate the task of parsing heterogeneous .rsp files into `parse_rsp`.

[the remark this answers is not in the document]
Comment (RyanTorok) [reply sw/host/cryptotest/testvectors/parsers/cryptotest_util.py:54]: #21728

[the remark this answers is not in the document]
Comment (sadmansakib-git) [reply sw/host/cryptotest/testvectors/parsers/cryptotest_util.py:54]: Had looked into the PR for refactoring the parser. LGTM. Thanks for the work!

As this PR is directly dependent on #21728 , I will push the relevant changes after it gets merged.

**Comment by milesdai** [comment]
@sadmansakib-git, #21728 has been merged with the updates to the `parse_rsp` function. Do you think you can rebase and try the new parser?

**Comment by sadmansakib-git** [comment]
> @sadmansakib-git, #21728 has been merged with the updates to the `parse_rsp` function. Do you think you can rebase and try the new parser?

@milesdai , I just tried the new parser after rebasing with my changes and pushed the new changes.

**Review by milesdai — APPROVED** [review]
LGTM! Thanks!

File: sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            "vendor": "nist",
+            "mode": "gcm",
+            "operation": args.operation.lower(),
+            "key_len": args.key_len,
+            "key": str_to_byte_array(test_vec["Key"]),
+            "aad": str_to_byte_array(test_vec["AAD"]),
+            "iv": str_to_byte_array(test_vec["IV"]),
+            "ciphertext": str_to_byte_array(test_vec["CT"]),
+            "plaintext": str_to_byte_array(test_vec["PT"]) if "PT" in test_vec else [],
+            "result": True if "PT" in test_vec else False,
```
Comment (RyanTorok) [inline sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py:33]: Why is the expected result "pass" for all decryption tests and "failure" for all encryption tests?

File: sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import sys
+import json
+import jsonschema
+
+from cryptotest_util import parse_rsp, str_to_byte_array
+
+
+def parse_testcases(args) -> None:
+    raw_testcases = parse_rsp(args.src)
+    testcases = list()
```
Comment (RyanTorok) [inline sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py:20]: Nit: rename variables to `test_cases` and `test_case`

**Comment by RyanTorok** [comment]
Derp, I saw a message from Miles to review this but only realized afterwards it was already merged.

File: sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py (hunk 3 above)
Comment (sadmansakib-git) [reply sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py:33]: Looks like a mistake from my side while working from another work setup. That's unfortunate that somehow it slipped through.
Currently, there are no FAIL cases in the GCM encryption test vectors. However, for decryption test vectors, if a test vector lacks a Plaintext (PT) field, the result of that particular test vector will be marked as FAIL.

Thanks a lot for pointing this out! I am sending a PR with the fix.

File: sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py (hunk 4 above)
Comment (sadmansakib-git) [reply sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py:20]: Noted

File: sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py (hunk 3 above)
Comment (sadmansakib-git) [reply sw/host/cryptotest/testvectors/parsers/nist_cavp_aes_gcm_parser.py:33]: #21881 21881

**Pull request merged**